tensorboard/plugins/image/summary.py

```diff
@@ -0,0 +1,135 @@
+                                tf.assert_type(images, tf.uint8),
+                                tf.assert_non_negative(max_outputs)]):
+    limited_images = images[:max_outputs]
+    encoded_images = tf.map_fn(tf.image.encode_png, limited_images,
+                               dtype=tf.string,
+                               name='encode_each_image')
+    image_shape = tf.shape(images)
+    dimensions = tf.stack([tf.as_string(image_shape[1], name='width'),
+                           tf.as_string(image_shape[2], name='height')],
+                          name='dimensions')
+    tensor = tf.concat([dimensions, encoded_images], axis=0)
+    return tf.summary.tensor_summary(name='image_summary',
+                                     tensor=tensor,
+                                     collections=collections,
+                                     summary_metadata=summary_metadata)
+
+
+def pb(name, images, max_outputs=3, display_name=None, description=None):
+  """Create an image summary protobuf.
+
+  This behaves as if you were to create an `op` with the same arguments
+  (wrapped with constant tensors where appropriate) and then execute
+  that summary op in a TensorFlow session.
+
+  Arguments:
+    name: A unique name for the generated summary, including any desired
+      name scopes.
+    images: An `np.array` representing pixel data with shape
+      `[k, w, h, c]`, where `k` is the number of images, `w` and `h` are
+      the width and height of the images, and `c` is the number of
+      channels, which should be 1, 3, or 4.
+    max_outputs: Optional `int`. At most this many images will be
+      emitted. If more than this many images are provided, the first
+      `max_outputs` many images will be used and the rest silently
+      discarded.
+    display_name: Optional name for this summary in TensorBoard, as a
+      `str`. Defaults to `name`.
+    description: Optional long-form description for this summary, as a
+      `str`. Markdown is supported. Defaults to empty.
+
+  Returns:
+    A `tf.Summary` protobuf object.
+  """
+  images = np.array(images).astype(np.uint8)
+  if images.ndim != 4:
+    raise ValueError('Shape %r must have rank 4' % (images.shape, ))
+
+  limited_images = images[:max_outputs]
+  encoded_images = [_encode_png(image) for image in limited_images]
+  (width, height) = (images.shape[1], images.shape[2])
+  content = [str(width), str(height)] + encoded_images
+  tensor = tf.make_tensor_proto(content, dtype=tf.string)
+
+  if display_name is None:
+    display_name = name
+  summary_metadata = metadata.create_summary_metadata(
+      display_name=display_name, description=description)
+
+  summary = tf.Summary()
+  summary.value.add(tag='%s/image_summary' % name,
+                    metadata=summary_metadata,
+                    tensor=tensor)
+  return summary
+
+
+def _encode_png(image):
+  # TODO(@wchargin): Pick a third-party PNG backend and implement this.
```

Comment:
Okay: I could do that in a separate PR, as discussed offline.

Once we add `pillow` support, under what circumstances would we expect `import pillow` to fail? I don't really understand the point of the `try`.

I agree with your assessment and also don't see any non-philosophical objections right now. We're not planning to remove `import tensorflow as tf` any time soon.

(Beholder doesn't use the summary system, right? So it can do as it pleases, using pillow or TensorFlow as it sees fit.)

> considering having TensorBase not store PNG encoded images, and instead store them as tensors

This would be a change later on, but I suspect that this may be impractical. PNG encoding is both expensive and effective. In my images demo, by far the slowest ops were PNG encoding—even though I was doing Θ(n²) 2D convolutions. [Here are the stats.](https://user-images.githubusercontent.com/4317806/28896053-c2312e98-778f-11e7-86e3-1d6541d1b151.png) Currently, image serving is essentially instant—a cache fetch—and so you can comfortably slide the "step" slider and watch the images update very quickly. This wouldn't be the case if the backend had to encode.

I suspect that this would also greatly increase the amount of data stored on disk. For instance, MRI scans are very large and mostly black, and should compress very well. This is even more true of audio—currently, we only use WAV, and the small-sized audio datasets of Googlers who work in this area are on the order of 10–20 GB. We don't want to see this kind of blowup for image demos, surely? (And we should also use OGG for audio…)